### ISSUE TITLE
Cannot import 'rti.connextdds' due to 'invalid syntax (rti.connextdds, line 18332)' (syntax-error)

### ISSUE THREAD

**Issue by jzenoz** [pylint-dev/pylint#9585]
### Bug description

Hi, tested using python3.10 and python3.11 with pylint version >=3.0.0a7 throws an error when trying to import 'rti.connextdds':

```python

$ python3.11 -m pylint --disable=missing-class-docstring,missing-module-docstring,missing-function-docstring,too-few-public-methods --verbose *
No config file found, using default configuration
************* Module hello_publisher
hello_publisher.py:2:0: E0001: Cannot import 'rti.connextdds' due to 'invalid syntax (rti.connextdds, line 18332)' (syntax-error)
hello_publisher.py:2:0: E0611: No name 'connextdds' in module 'rti' (no-name-in-module)
************* Module hello_subscriber
hello_subscriber.py:1:0: E0001: Cannot import 'rti.connextdds' due to 'invalid syntax (rti.connextdds, line 18332)' (syntax-error)
hello_subscriber.py:1:0: E0611: No name 'connextdds' in module 'rti' (no-name-in-module)

---------------------------------------------------------------------------------------------------
Your code has been rated at 0.91/10 (previous run: 0.45/10, +0.45)
Checked 3 files, skipped 3 files
```

```python
$ python3.11 -m pip freeze | grep rti
rti.connext==7.3.0
```

Using their hello world setup described here: https://community.rti.com/static/documentation/connext-dds/7.3.0/doc/api/connext_dds/api_python/hello.html

```python
$ cat hello.py 
import rti.types as idl

@idl.struct
class HelloWorld:
    message: str = ""
```
```python
$ cat hello_publisher.py 
import time
import rti.connextdds as dds
from hello import HelloWorld

participant = dds.DomainParticipant(domain_id=0)
topic = dds.Topic(participant, 'HelloWorld Topic', HelloWorld)
writer = dds.DataWriter(participant.implicit_publisher, topic)

for i in range(10):
    writer.write(HelloWorld(message=f'Hello World! #{i}'))
    time.sleep(1)
```
```python
$ cat hello_subscriber.py 
import rti.connextdds as dds
import rti.asyncio
from hello import HelloWorld

participant = dds.DomainParticipant(domain_id=0)
topic = dds.Topic(participant, 'HelloWorld Topic', HelloWorld)
reader = dds.DataReader(participant.implicit_subscriber, topic)

async def print_data():
    async for data in reader.take_data_async():
        print(f"Received: {data}")

rti.asyncio.run(print_data())
```

```shell
$ python3.11 -m pip install pylint==2.17.7
...
Successfully installed pylint-2.17.7
```
```python
$ python3.11 -m pylint --disable=missing-class-docstring,missing-module-docstring,missing-function-docstring,too-few-public-methods --verbose *
No config file found, using default configuration
************* Module hello_publisher
hello_publisher.py:5:14: I1101: Module 'rti.connextdds' has no 'DomainParticipant' member, but source is unavailable. Consider adding this module to extension-pkg-allow-list if you want to perform analysis based on run-time introspection of living objects. (c-extension-no-member)
hello_publisher.py:6:8: I1101: Module 'rti.connextdds' has no 'Topic' member, but source is unavailable. Consider adding this module to extension-pkg-allow-list if you want to perform analysis based on run-time introspection of living objects. (c-extension-no-member)
hello_publisher.py:7:9: I1101: Module 'rti.connextdds' has no 'DataWriter' member, but source is unavailable. Consider adding this module to extension-pkg-allow-list if you want to perform analysis based on run-time introspection of living objects. (c-extension-no-member)
************* Module hello_subscriber
hello_subscriber.py:5:14: I1101: Module 'rti.connextdds' has no 'DomainParticipant' member, but source is unavailable. Consider adding this module to extension-pkg-allow-list if you want to perform analysis based on run-time introspection of living objects. (c-extension-no-member)
hello_subscriber.py:6:8: I1101: Module 'rti.connextdds' has no 'Topic' member, but source is unavailable. Consider adding this module to extension-pkg-allow-list if you want to perform analysis based on run-time introspection of living objects. (c-extension-no-member)
hello_subscriber.py:7:9: I1101: Module 'rti.connextdds' has no 'DataReader' member, but source is unavailable. Consider adding this module to extension-pkg-allow-list if you want to perform analysis based on run-time introspection of living objects. (c-extension-no-member)

--------------------------------------------------------------------
Your code has been rated at 10.00/10 (previous run: 10.00/10, +0.00)
```

### Command used

```shell
python3.11 -m pylint --disable=missing-class-docstring,missing-module-docstring,missing-function-docstring,too-few-public-methods --verbose *
```

### Pylint output

```python
No config file found, using default configuration
************* Module hello_publisher
hello_publisher.py:2:0: E0001: Cannot import 'rti.connextdds' due to 'invalid syntax (rti.connextdds, line 18332)' (syntax-error)
hello_publisher.py:2:0: E0611: No name 'connextdds' in module 'rti' (no-name-in-module)
************* Module hello_subscriber
hello_subscriber.py:1:0: E0001: Cannot import 'rti.connextdds' due to 'invalid syntax (rti.connextdds, line 18332)' (syntax-error)
hello_subscriber.py:1:0: E0611: No name 'connextdds' in module 'rti' (no-name-in-module)

---------------------------------------------------------------------------------------------------
Your code has been rated at 0.91/10 (previous run: 0.45/10, +0.45)
Checked 3 files, skipped 3 files
```

### Expected behavior

```python
No config file found, using default configuration
************* Module hello_publisher
hello_publisher.py:5:14: I1101: Module 'rti.connextdds' has no 'DomainParticipant' member, but source is unavailable. Consider adding this module to extension-pkg-allow-list if you want to perform analysis based on run-time introspection of living objects. (c-extension-no-member)
hello_publisher.py:6:8: I1101: Module 'rti.connextdds' has no 'Topic' member, but source is unavailable. Consider adding this module to extension-pkg-allow-list if you want to perform analysis based on run-time introspection of living objects. (c-extension-no-member)
hello_publisher.py:7:9: I1101: Module 'rti.connextdds' has no 'DataWriter' member, but source is unavailable. Consider adding this module to extension-pkg-allow-list if you want to perform analysis based on run-time introspection of living objects. (c-extension-no-member)
************* Module hello_subscriber
hello_subscriber.py:5:14: I1101: Module 'rti.connextdds' has no 'DomainParticipant' member, but source is unavailable. Consider adding this module to extension-pkg-allow-list if you want to perform analysis based on run-time introspection of living objects. (c-extension-no-member)
hello_subscriber.py:6:8: I1101: Module 'rti.connextdds' has no 'Topic' member, but source is unavailable. Consider adding this module to extension-pkg-allow-list if you want to perform analysis based on run-time introspection of living objects. (c-extension-no-member)
hello_subscriber.py:7:9: I1101: Module 'rti.connextdds' has no 'DataReader' member, but source is unavailable. Consider adding this module to extension-pkg-allow-list if you want to perform analysis based on run-time introspection of living objects. (c-extension-no-member)
```

### Pylint version

```shell
pylint 3.1.0
astroid 3.1.0
Python 3.11.0rc1 (main, Aug 12 2022, 10:02:14) [GCC 11.2.0]

pylint==3.0.0a6 works fine, as does pylint==2.17.7 and presumably versions before
```

### OS / Environment

```shell
$ uname -a
Linux WA-ZB-5060 5.15.146.1-microsoft-standard-WSL2 #1 SMP Thu Jan 11 04:09:03 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux
```
```shell
$ cat /etc/*release*
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=22.04
DISTRIB_CODENAME=jammy
DISTRIB_DESCRIPTION="Ubuntu 22.04.4 LTS"
PRETTY_NAME="Ubuntu 22.04.4 LTS"
NAME="Ubuntu"
VERSION_ID="22.04"
VERSION="22.04.4 LTS (Jammy Jellyfish)"
VERSION_CODENAME=jammy
ID=ubuntu
ID_LIKE=debian
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
UBUNTU_CODENAME=jammy
```

### Additional dependencies

```python
rti.connext==7.3.0
```

**Comment by UlrichEckhardt** [comment]
Note up front: I'm not a maintainer of pylint, so take the following with a grain of salt.

Some observations:
- Python 3.11.0rc1 is a release candidate ("rc"), not something you should use for production. I'd suggest retrying with an actual release (current is 3.11.9) instead.
- Current Python version in Ubuntu 22.04 is 3.10.12, as far as I can tell. I tried to reproduce your setup with Docker though, so there may be some differences. How did you install those Python versions?
- You mention pytest 3, but that's five(!) major versions behind. Not sure about their release naming, but 3.0.0a7 also seems like a prerelease version. Did you mean pylint there?
- Your commandline uses `*` as an argument, which is replaced by the shell with every filename in the current working directory. That's surely not intended, though it might work by chance. It also makes it non-reproducible without knowing your whole directory content.
- You say you installed pylint 2.17.7, but lateron say that 3.1.0 is installed. This makes it difficult to understand what you actually did and what happened.

I tried the following in an empty dir on a Ubuntu 22.04 system:
```shell
python3 -m venv pyenv
source pyenv/bin/activate.fish
python -m pip install pylint==2.17.7 # or 3.1.0, or 3.0.0, doesn't matter
python -m pip install rti.connext==7.3.0
# create a hello.py as per your instructions above
python -m pylint --disable=missing-class-docstring,missing-module-docstring,missing-function-docstring,too-few-public-methods --verbose hello.py
```
With that, I couldn't reproduce what you wrote above. Can you provide step-by-step instructions?

**Comment by jzenoz** [comment]
> Note up front: I'm not a maintainer of pylint, so take the following with a grain of salt.
> 
> Some observations:
> 
> * Python 3.11.0rc1 is a release candidate ("rc"), not something you should use for production. I'd suggest retrying with an actual release (current is 3.11.9) instead.
> * Current Python version in Ubuntu 22.04 is 3.10.12, as far as I can tell. I tried to reproduce your setup with Docker though, so there may be some differences. How did you install those Python versions?
> * You mention pytest 3, but that's five(!) major versions behind. Not sure about their release naming, but 3.0.0a7 also seems like a prerelease version. Did you mean pylint there?
> * Your commandline uses `*` as an argument, which is replaced by the shell with every filename in the current working directory. That's surely not intended, though it might work by chance. It also makes it non-reproducible without knowing your whole directory content.
> * You say you installed pylint 2.17.7, but lateron say that 3.1.0 is installed. This makes it difficult to understand what you actually did and what happened.
> 
> I tried the following in an empty dir on a Ubuntu 22.04 system:
> 
> ```shell
> python3 -m venv pyenv
> source pyenv/bin/activate.fish
> python -m pip install pylint==2.17.7 # or 3.1.0, or 3.0.0, doesn't matter
> python -m pip install rti.connext==7.3.0
> # create a hello.py as per your instructions above
> python -m pylint --disable=missing-class-docstring,missing-module-docstring,missing-function-docstring,too-few-public-methods --verbose hello.py
> ```
> 
> With that, I couldn't reproduce what you wrote above. Can you provide step-by-step instructions?
@UlrichEckhardt 
* I tested using python3.10 as well, as stated in my post, the default Ubuntu 22.04 python 3.10.12, the 3.11 was my 2nd set of tests
* the python3.11 installed via `apt install python3.11`, 3.10 default
* Aye, a typo, fixed!
* directory content is solely the rti dds hello world scripts that I cat
* I tested using both pylint 2.17.7 and pytest 3.1.0, pytest 2.17.7 works, pytest 3.1.0 does not.

Fresh reproducer on Windows 11:
```bash
wsl --install -d Ubuntu
wsl
# Welcome to Ubuntu 22.04.3 LTS (GNU/Linux 5.10.102.1-microsoft-standard-WSL2 x86_64)
apt update
apt install python3-pip
# pip 22.0.2 from /usr/lib/python3/dist-packages/pip (python 3.10)
pip install pylint
pylint --version
# pylint 3.1.0
# astroid 3.1.0
# Python 3.10.12 (main, Nov 20 2023, 15:14:05) [GCC 11.4.0]
pip install rti.connext
# rti.connext-7.3.0

mkdir test
cd test

echo 'import rti.types as idl

@idl.struct
class HelloWorld:
    message: str = ""' > hello.py

echo "import time
import rti.connextdds as dds
from hello import HelloWorld

participant = dds.DomainParticipant(domain_id=0)
topic = dds.Topic(participant, 'HelloWorld Topic', HelloWorld)
writer = dds.DataWriter(participant.implicit_publisher, topic)

for i in range(10):
    writer.write(HelloWorld(message=f'Hello World! #{i}'))
    time.sleep(1)" > hello_publisher.py

echo 'import rti.connextdds as dds
import rti.asyncio
from hello import HelloWorld

participant = dds.DomainParticipant(domain_id=0)
topic = dds.Topic(participant, "HelloWorld Topic", HelloWorld)
reader = dds.DataReader(participant.implicit_subscriber, topic)

async def print_data():
    async for data in reader.take_data_async():
        print(f"Received: {data}")

rti.asyncio.run(print_data())' > hello_subscriber.py

python3 -m pylint --disable=missing-class-docstring,missing-module-docstring,missing-function-docstring,too-few-public-methods --verbose hello.py hello_publisher.py hello_subscriber.py
No config file found, using default configuration
************* Module hello_publisher
hello_publisher.py:2:0: E0001: Cannot import 'rti.connextdds' due to 'invalid syntax (rti.connextdds, line 18332)' (syntax-error)
hello_publisher.py:2:0: E0611: No name 'connextdds' in module 'rti' (no-name-in-module)
************* Module hello_subscriber
hello_subscriber.py:1:0: E0001: Cannot import 'rti.connextdds' due to 'invalid syntax (rti.connextdds, line 18332)' (syntax-error)
hello_subscriber.py:1:0: E0611: No name 'connextdds' in module 'rti' (no-name-in-module)

--------------------------------------------------------------------
Your code has been rated at 0.91/10
Checked 3 files, skipped 3 files
---------------------------------------------------------------------------------------------------
Your code has been rated at 0.00/10 (previous run: 0.00/10, +0.00)
Checked 2 files, skipped 2 files
```

**Comment by jpsamper2009** [comment]
@UlrichEckhardt @jzenoz Did you ever resolve this issue? 

I'm seeing it too on Ubuntu 24.04 with Python 3.12. It seems like the bug is in `astroid` (https://github.com/pylint-dev/astroid/issues/2640)

**Comment by jacobtylerwalls** [comment]
Hi, thanks for the report. We will need the python code found at this line of this module to be able to evaluate this any further:

```
rti.connextdds, line 18332
```

Please reopen if you can provide that. Thanks!

**Comment by jpsamper2009** [comment]
@jacobtylerwalls Ok, so I did some digging. TL;DR there is a syntax error.

For completeness:

The syntax error appears to be in a file `rti/connextdds/__init.pyi__` file that gets generated. 

The source code is in https://github.com/rticommunity/connextdds-py, and the file can be generated by following https://github.com/rticommunity/connextdds-py/blob/master/docs/source/building.rst.

**Comment by UlrichEckhardt** [comment]
FWIW, I just tried to reproduce this again, using Ubuntu 22.04, using the [instructions here](https://github.com/pylint-dev/pylint/issues/9585#issuecomment-2089417218), only using a Docker container instead of WSL. A few versions changed and the pylint call in the end didn't fail. Inspecting the connextdds files installed in the venv, I also located the `ValuesView` class, which seemed to be the culprit, and it seems to have changed:
```python
class ValuesView():
    def __iter__(self) -> typing.Iterator: ...
    def __len__(self) -> int: ...
    pass
```
rti.connext is installed in version 7.5.0.


In summary, I guess this really was just a bug in some other library or utility, though I must say that it's just a guess.

**Comment by Pierre-Sassoulas** [comment]
I'm going to close as no longer reproduceable / probably not an actual problem, I skimmed the issue but it seems there was an actual syntax error in a dependency, and it's expected for pylint to fail on actual syntax error.

```python
 class ItemsView[StringMap]():
     def __iter__(self) -> typing.Iterator: ...
     def __len__(self) -> int: ...
     pass
```
The ``pass``s a SyntaxError, should be removed, or indented to the right.

**Comment by UlrichEckhardt** [comment]
No, the `pass` is not a syntax error and should not be indented. Doing so won't fix above code for Python 3.8 (which is one version I have installed locally, in addition to 3.12).

Try this for example:
```python
import typing

class Foo():
    def __iter__(self) -> typing.Iterator: ...
    pass
```
Yes, the `pass` is dead code, but harmless. Pythons 3.8 and 3.12 accept this.

Try this for comparison:
```python
import typing

class Bar:
    pass

class Foo[Bar]():
    def __iter__(self) -> typing.Iterator: ...
    pass
```
Here, only Python 3.12 accepts this, 3.8 rejects it:
```none
  File "x.py", line 6
    class Foo[Bar]():
             ^
SyntaxError: invalid syntax
```

BTW: I have no idea what the `[Bar]` syntax means, to be honest.

**Comment by jzenoz** [comment]
Rti.connext 7.5.0 isn't LTS so we can't use it yet but it is hopeful! Thanks for checking again. Also that syntax is defined for Python 3.12 PEP 695:
https://peps.python.org/pep-0695/

**Comment by Pierre-Sassoulas** [comment]
My bad I just assumed and didn't check. It's still impossible to reproduce though.

**Comment by jzenoz** [comment]
So I've left that project that was hung on this error but it still should
be resolved either root cause wise or otherwise. Does the reproducer above
not work for anyone?

**Comment by Pierre-Sassoulas** [comment]
I can reproduce on pylint main with:

```bash
uv venv --python 3.11 /tmp/venv311
uv pip install --python /tmp/venv311/bin/python -e . # from pylint dev dir obviousely
uv pip install --python /tmp/venv311/bin/python "rti.connext==7.3.0"
mkdir -p /tmp/9585 && cd /tmp/9585
cat > test.py <<'EOF'
import rti.connextdds as dds
participant = dds.DomainParticipant(domain_id=0)
EOF
/tmp/venv311/bin/python -m pylint test.py
```

Root cause: astroid 3.x added .pyi to PY_SOURCE_EXTS, so file_from_modpath(['rti', 'connextdds']) now returns the stub path instead of the .so.
